### ISSUE TITLE
no-useless-return has a false negative with try/catch

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7481]
**Tell us about your environment**
- **ESLint Version:** 3.9.0
- **Node Version:** 6.9.1
- **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

``` yaml
rules:
  no-useless-return: error
```

**What did you do? Please include the actual source code causing the issue.**

``` js

function foo() {
  try {
    return;
  } catch (err) {
    foo();
  }
}

```

**What did you expect to happen?**

This should report an error, because the `return` statement can be safely removed.

**What actually happened? Please include the actual, raw output from ESLint.**

No error.

**Comment by mysticatea** [comment]
Thank you for this issue.

In fact, this is caused by the current code path analyzer.
Since the analyzer cannot know the places which throw exceptions, it generates 2 code paths which go to the `catch` block from both the first statement and the last statement.
But in this case, the `return;` statement throws no exception clearly. I will investigate to improve the code path.

**Comment by not-an-aardvark** [comment]
When linting this code:

```js
try { a; b; c; } catch (e) {}
```

I got this code path:

<details>
<summary>DOT</summary>

```dot
digraph {
node[shape=box,style="rounded,filled",fillcolor=white];
initial[label="",shape=circle,style=filled,fillcolor=black,width=0.25,height=0.25];
final[label="",shape=doublecircle,style=filled,fillcolor=black,width=0.25,height=0.25];
s1_1[label="Program\nTryStatement\nBlockStatement\nExpressionStatement\nIdentifier (a)\nIdentifier:exit (a)"];
s1_2[label="ExpressionStatement\nIdentifier (b)\nExpressionStatement\nIdentifier (c)\nExpressionStatement:exit\nIdentifier:exit (b)\nExpressionStatement:exit\nIdentifier:exit (c)\nExpressionStatement:exit\nBlockStatement:exit"];
s1_4[label="TryStatement:exit\nProgram:exit"];
s1_3[label="CatchClause\nIdentifier (e)\nBlockStatement\nIdentifier:exit (e)\nBlockStatement:exit\nCatchClause:exit"];
initial->s1_1->s1_2->s1_4;
s1_1->s1_3->s1_4->final;
}
```
</details>

<details>
<summary>Image</summary>

![](https://i.gyazo.com/87a6d50149dd81a6fbfc6e806e50e7e3.png)
</details>

---

But I expected to get something like this:

<details>
<summary>DOT</summary>

```dot
digraph {
node[shape=box,style="rounded,filled",fillcolor=white];
initial[label="",shape=circle,style=filled,fillcolor=black,width=0.25,height=0.25];
final[label="",shape=doublecircle,style=filled,fillcolor=black,width=0.25,height=0.25];
s1_1[label="Program\nTryStatement\nBlockStatement\nExpressionStatement\nIdentifier (a)"];
s1_2[label="Identifier:exit (a)\nExpressionStatement:exit\nExpressionStatement\nIdentifier (b)"];
s1_3[label="Identifier:exit (b)\nExpressionStatement:exit\nExpressionStatement\nIdentifier (c)"];
s1_4[label="Identifier:exit (c)\nExpressionStatement:exit\nBlockStatement:exit"];
s1_5[label="CatchClause\nIdentifier (e)\nBlockStatement\nIdentifier:exit (e)\nBlockStatement:exit\nCatchClause:exit"];
s1_6[label="TryStatement:exit\nProgram:exit"];
initial->s1_1->s1_2->s1_3->s1_4->s1_6;
s1_1->s1_5->s1_6->final;
s1_2->s1_5;
s1_3->s1_5;
}
```
</details>

<details>
<summary>Image</summary>

![](https://i.gyazo.com/3b0feae93cc73e147d32bfa27b384c19.png)
</details>

---

Is it intentional that each `try` statement only contains one error point (and one exit point at the end), rather than having an exit point for every possible error?

**Comment by mysticatea** [comment]
Yes, it's intentional because I thought that making every throwable path is too many. https://github.com/eslint/eslint/pull/3559#issuecomment-136100166

I'm sorry for I have not done this.

**Comment by not-an-aardvark** [comment]
I see, that makes sense.

* Do you think it would still be too many paths if we used scope analysis to avoid impossible `ReferenceError`s? (It might still be complex because we would have to keep track of TDZ references).

As a smaller fix for this issue, maybe we could avoid adding the edge to the `CatchClause` if the `TryStatement` has no throwable paths.

**Comment by mysticatea** [comment]
It makes sense if we avoid making throwable path at references which never throw.
If we make every throwable path, I think we should measure impact to performance.

**Comment by not-an-aardvark** [comment]
Actually, I suppose the second solution wouldn't resolve this issue because something like this would still report an error:

```js
try {
  foo();
  return;
} catch (err) {
  foo();
}
```

**Comment by nzakas** [comment]
@mysticatea @not-an-aardvark are either of you interested in addressing this?

**Comment by guilhermejcgois** [comment]
I have free time until 1st december, can I go for it @mysticatea ? Any tip will be very appreciate 😄 

> Taking a look into this issue seems that in the current version we still have it, adding the example of @not-an-aardvark and https://github.com/eslint/eslint/blob/261d7287820253408ec87c344beccdba2fe829a4/tests/lib/rules/no-useless-return.js#L308-L329 cause the test to fail

**Comment by nzakas** [comment]
@guilhermejcgois they have long since left the project. You are welcome to pick this up.

**Comment by ljharb** [comment]
This has an open PR, #16693

**Comment by sam3k** [comment]
Update:

- This might be the actual opened PR? [#16593](https://github.com/eslint/eslint/pull/16593) 
- We're waiting to hear from @guilhermejcgois on one last edge case that needs to be addressed in the PR

**Comment by nzakas** [comment]
Looks like we will need help finishing up the PR for this.
